File: backend/vexel-rs/src/overlaps.rs
//! Overlap decomposition: blended regions become overlapping shapes.
//!
//! Where a semi-transparent shape T lies over another shape X, the traced image
//! has an extra region C whose colour is a blend: `C = α·T + (1−α)·X`. A
//! designer drew T and X, not C. Every region C explained this way by a pair of
//! its neighbours is absorbed into *both* shapes: T (painted on top with
//! `fill-opacity α`) and X (beneath). Absorption is transitive. A top shape only
//! keeps its absorptions when the extended outline is *simpler* than the bare
//! region.

use crate::core::contours::{find_contours, pad};
use crate::core::grid::{Grid, Mask};
use crate::core::labels::{self, Labels};
use crate::core::morphology::convex_hull_area;
use crate::curves::{fit_shape, CurveParams, Shape, P};
use crate::fills::Fill;
use crate::merge::adjacency;
use std::collections::{HashMap, HashSet};

#[derive(Default)]
pub struct Decomposition {
    /// extended footprint per shape label
    pub masks: HashMap<i32, Mask>,
    /// top shapes: true colour + opacity
    pub fills: HashMap<i32, Fill>,
    /// blended regions that are no longer drawn
    pub removed: HashSet<i32>,
    /// (top, below) ordering constraints
    pub above: Vec<(i32, i32)>,
}

impl Decomposition {
    pub fn empty(&self) -> bool {
        self.removed.is_empty()
    }
}

fn solidity(mask: &Mask) -> f64 {
    let area = mask.count() as f64;
    if area == 0.0 {
        return 0.0;
    }
    area / convex_hull_area(mask)
}

/// The union outline is a primitive, or clearly more convex than the part.
fn simpler(union: &Mask, part: &Mask, curve_params: &CurveParams) -> bool {
    let field = Grid { h: union.h, w: union.w, data: union.data.iter().map(|b| *b as u8 as f64).collect() };
    let polys: Vec<Vec<P>> = find_contours(&pad(&field, 0.0), 0.5)
        .into_iter()
        .filter(|c| c.len() >= 8)
        .map(|c| c.iter().map(|(r, cc)| [cc - 0.5, r - 0.5]).collect())
        .collect();
    if polys.len() == 1 && !matches!(fit_shape(&polys, curve_params), Shape::Path { .. }) {
        return true;
    }
    solidity(union) > solidity(part) + 0.06
}

fn dot3(a: &[f64], b: &[f64]) -> f64 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}

fn dist3(a: &[f64], b: &[f64]) -> f64 {
    ((a[0] - b[0]).powi(2) + (a[1] - b[1]).powi(2) + (a[2] - b[2]).powi(2)).sqrt()
}

/// Solve `C = α·Tc + (1−α)·X` for (α, Tc, rms residual); colours rgba 0–255.
fn blend(t_vis: &[f64; 4], c_vis: &[f64; 4], x_vis: &[f64; 4], bg: Option<&[f64; 4]>) -> Option<(f64, [f64; 3], f64)> {
    let (alpha, tc): (f64, [f64; 3]);
    if t_vis[3] < 250.0 {
        // semi-transparent over transparency: alpha is the opacity
        let a = t_vis[3] / 255.0;
        if !(0.12..=0.95).contains(&a) {
            return None; // an almost-opaque shape does not produce a visible blend
        }
        alpha = a;
        tc = [t_vis[0], t_vis[1], t_vis[2]];
    } else {
        let bg = bg?;
        let d1 = [t_vis[0] - c_vis[0], t_vis[1] - c_vis[1], t_vis[2] - c_vis[2]];
        let d2 = [bg[0] - x_vis[0], bg[1] - x_vis[1], bg[2] - x_vis[2]];
        let den = dot3(&d2, &d2);
        if den < 1e-6 {
            return None;
        }
        let one_minus = dot3(&d1, &d2) / den;
        let a = 1.0 - one_minus;
        if !(0.12..=0.97).contains(&a) {
            return None;
        }
        let mut t = [0.0f64; 3];
        for i in 0..3 {
            t[i] = (t_vis[i] - one_minus * bg[i]) / a;
        }
        if t.iter().any(|v| *v < -12.0 || *v > 267.0) {
            return None;
        }
        alpha = a;
        tc = [t[0].clamp(0.0, 255.0), t[1].clamp(0.0, 255.0), t[2].clamp(0.0, 255.0)];
    }
    let mut resid = 0.0;
    for i in 0..3 {
        let pred = alpha * tc[i] + (1.0 - alpha) * x_vis[i];
        resid += (pred - c_vis[i]).powi(2);
    }
    let resid = (resid / 3.0).sqrt();
    // The blend must be a *different* colour from both parents; otherwise C is
    // just an anti-aliased sliver of one of them, not an overlap.
    if dist3(c_vis, t_vis) < 6.0 || dist3(c_vis, x_vis) < 6.0 {
        return None;
    }
    Some((alpha, tc, resid))
}

pub fn decompose_overlaps(
    l: &Labels,
    fills: &HashMap<i32, Fill>,
    visible: &HashMap<i32, bool>,
    curve_params: &CurveParams,
    tol: f64,
) -> Decomposition {
    const MAX_REGION_FRAC: f64 = 0.6;
    let mut dec = Decomposition::default();
    let edges = adjacency(l, None);
    let mut nbrs: HashMap<i32, HashSet<i32>> = HashMap::new();
    for (a, b) in edges.keys() {
        nbrs.entry(*a).or_default().insert(*b);
        nbrs.entry(*b).or_default().insert(*a);
    }
    let solid_rgba = |lab: i32| -> Option<[f64; 4]> {
        match fills.get(&lab) {
            Some(Fill::Solid { rgba }) => Some(*rgba),
            _ => None,
        }
    };
    let mut solid_vis: Vec<i32> = fills
        .keys()
        .copied()
        .filter(|lab| visible.get(lab).copied().unwrap_or(false) && solid_rgba(*lab).is_some())
        .collect();
    solid_vis.sort_unstable();
    if solid_vis.len() < 3 {
        return dec;
    }
    let solid_set: HashSet<i32> = solid_vis.iter().copied().collect();

    let counts = labels::bincount(l);
    let areas: HashMap<i32, usize> = solid_vis.iter().map(|lab| (*lab, counts[*lab as usize])).collect();

    let (h, w) = (l.h, l.w);
    let mut border_labels: HashSet<i32> = HashSet::new();
    for c in 0..w {
        border_labels.insert(l.data[c]);
        border_labels.insert(l.data[(h - 1) * w + c]);
    }
    for r in 0..h {
        border_labels.insert(l.data[r * w]);
        border_labels.insert(l.data[r * w + w - 1]);
    }
    let background = solid_vis
        .iter()
        .filter(|lab| border_labels.contains(lab))
        .max_by_key(|lab| areas[lab])
        .copied();
    let bg_colour = background.and_then(|b| {
        let c = solid_rgba(b)?;
        if c[3] > 250.0 {
            Some(c)
        } else {
            None
        }
    });
    let total = l.len();

    // 1. explain each candidate region by the best (top, under) pair of its neighbours
    let mut explained: HashMap<i32, (i32, i32, f64, [f64; 3])> = HashMap::new();
    for c in &solid_vis {
        if Some(*c) == background || areas[c] as f64 > MAX_REGION_FRAC * total as f64 {
            continue;
        }
        let mut cand: Vec<i32> = nbrs
            .get(c)
            .map(|s| s.iter().copied().filter(|n| solid_set.contains(n) && Some(*n) != background).collect())
            .unwrap_or_default();
        cand.sort_unstable();
        let mut best: Option<(f64, i32, i32, f64, [f64; 3])> = None;
        for t in &cand {
            for x in &cand {
                if t == x {
                    continue;
                }
                let Some(res) = blend(&solid_rgba(*t).unwrap(), &solid_rgba(*c).unwrap(), &solid_rgba(*x).unwrap(), bg_colour.as_ref())
                else {
                    continue;
                };
                if res.2 > tol {
                    continue;
                }
                if best.as_ref().is_none_or(|b| res.2 < b.0) {
                    best = Some((res.2, *t, *x, res.0, res.1));
                }
            }
        }
        if let Some((_, t, x, a, tc)) = best {
            explained.insert(*c, (t, x, a, tc));
        }
    }
    if explained.is_empty() {
        return dec;
    }

    // 2. a top shape keeps its absorptions only if the extended outline is simpler
    let mut tops: HashMap<i32, Vec<i32>> = HashMap::new();
    let mut expl_keys: Vec<i32> = explained.keys().copied().collect();
    expl_keys.sort_unstable();
    for c in &expl_keys {
        tops.entry(explained[c].0).or_default().push(*c);
    }
    let mut accepted: HashMap<i32, (i32, i32, f64, [f64; 3])> = HashMap::new();
    let mut top_keys: Vec<i32> = tops.keys().copied().collect();
    top_keys.sort_unstable();
    for t in &top_keys {
        let cs = &tops[t];
        let t_mask = labels::mask_of(l, *t);
        let mut union = t_mask.clone();
        for i in 0..union.len() {
            if cs.contains(&l.data[i]) {
                union.data[i] = true;
            }
        }
        if simpler(&union, &t_mask, curve_params) {
            for c in cs {
                accepted.insert(*c, explained[c]);
            }
        }
    }
    if accepted.is_empty() {
        return dec;
    }

    // 3. footprints: T ∪ C for tops, X ∪ C beneath, transitively through removed regions
    let mut members: HashMap<i32, HashSet<i32>> = HashMap::new();
    fn add(
        accepted: &HashMap<i32, (i32, i32, f64, [f64; 3])>,
        members: &mut HashMap<i32, HashSet<i32>>,
        shape: i32,
        region: i32,
        depth: usize,
    ) {
        if let Some((t, x, _, _)) = accepted.get(&shape) {
            if depth < 8 {
                // the shape is itself a removed overlap: pass it to its owners
                add(accepted, members, *t, region, depth + 1);
                add(accepted, members, *x, region, depth + 1);
                return;
            }
        }
        members.entry(shape).or_default().insert(region);
    }

    let mut acc_keys: Vec<i32> = accepted.keys().copied().collect();
    acc_keys.sort_unstable();
    for c in &acc_keys {
        let (t, x, _, _) = accepted[c];
        add(&accepted, &mut members, t, *c, 0);
        add(&accepted, &mut members, x, *c, 0);
        dec.removed.insert(*c);
    }

    /// Follow a removed region to the shape that owns it (0 = top, 1 = under); cycle-safe.
    fn resolve(accepted: &HashMap<i32, (i32, i32, f64, [f64; 3])>, mut shape: i32, index: usize) -> i32 {
        let mut seen: HashSet<i32> = HashSet::new();
        seen.insert(shape);
        while let Some(e) = accepted.get(&shape) {
            shape = if index == 0 { e.0 } else { e.1 };
            if !seen.insert(shape) {
                break;
            }
        }
        shape
    }

    for c in &acc_keys {
        let (t, x, alpha, tc) = accepted[c];
        if !accepted.contains_key(&t) {
            dec.fills.insert(t, Fill::Solid { rgba: [tc[0], tc[1], tc[2], alpha * 255.0] });
        }
        let top_owner = resolve(&accepted, t, 0);
        let under = resolve(&accepted, x, 1);
        if top_owner != under && !accepted.contains_key(&top_owner) && !accepted.contains_key(&under) {
            dec.above.push((top_owner, under));
        }
    }

    let mut shape_keys: Vec<i32> = members.keys().copied().collect();
    shape_keys.sort_unstable();
    for shape in shape_keys {
        if dec.removed.contains(&shape) {
            continue;
        }
        let regs = &members[&shape];
        let mask = Grid {
            h,
            w,
            data: l.data.iter().map(|v| *v == shape || regs.contains(v)).collect(),
        };
        dec.masks.insert(shape, mask);
    }
    dec
}

File: backend/studi0trace/engines/vexel/overlaps.py
"""Overlap decomposition: blended regions become overlapping shapes.

Where a semi-transparent shape T lies over another shape X, the traced image
has an extra region C whose colour is a blend: `C = α·T_colour + (1−α)·X_vis`.
A designer drew T and X, not C. Every region C that is explained this way by a
pair of its neighbours is absorbed into *both* shapes: T (painted on top with
`fill-opacity α`) and X (beneath). Absorption is transitive — the triple
overlap of three circles is "T over (A∩B)", and A∩B belongs to A and B, so the
triple joins A and B as well. A top shape only keeps its absorptions when the
extended outline is *simpler* than the bare region (a primitive, or clearly
more convex — the T-junction cue that the occluded contour continues).

Against a transparent background a shape's own alpha is its opacity; over an
opaque background the two blend equations give α and the true colour.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from skimage.measure import find_contours
from skimage.morphology import convex_hull_image

from studi0trace.engines.vexel.curves import CurveParams, PathShape, fit_shape
from studi0trace.engines.vexel.fills import Fill, Solid
from studi0trace.engines.vexel.merge import adjacency


# Least share of an overlap region's outline that must run along the two shapes
# it is the overlap of (or along other overlaps): see `decompose_overlaps`.
OVERLAP_OUTLINE = 0.5


@dataclass
class Decomposition:
    masks: dict[int, np.ndarray] = field(default_factory=dict)  # extended footprint per shape label
    fills: dict[int, Solid] = field(default_factory=dict)  # top shapes: true colour + opacity
    removed: set[int] = field(default_factory=set)  # blended regions that are no longer drawn
    above: list[tuple[int, int]] = field(default_factory=list)  # (top, below) ordering constraints
    over_backdrop: set[int] = field(default_factory=set)  # opaque regions made tops, solved over the opaque backdrop

    @property
    def empty(self) -> bool:
        return not self.removed


def _solidity(mask: np.ndarray) -> float:
    area = float(mask.sum())
    return area / float(convex_hull_image(mask).sum()) if area else 0.0


def _simpler(union: np.ndarray, part: np.ndarray, curve_params: CurveParams) -> bool:
    """The union outline is a primitive, or clearly more convex than the part."""
    polys = [np.column_stack([c[:, 1] - 0.5, c[:, 0] - 0.5]) for c in find_contours(np.pad(union.astype(np.float32), 1), 0.5)]
    polys = [p for p in polys if len(p) >= 8]
    if len(polys) == 1 and not isinstance(fit_shape(polys, curve_params), PathShape):
        return True
    return _solidity(union) > _solidity(part) + 0.06


def _blend(t_vis: np.ndarray, c_vis: np.ndarray, x_vis: np.ndarray, bg: np.ndarray | None) -> tuple[float, np.ndarray, float] | None:
    """Solve C = α·Tc + (1−α)·X for (α, Tc, rms residual); colours rgba 0–255."""
    if t_vis[3] < 250:  # semi-transparent over transparency: alpha is the opacity
        alpha = float(t_vis[3] / 255.0)
        if not 0.12 <= alpha <= 0.95:
            return None  # an almost-opaque shape does not produce a visible blend
        tc = t_vis[:3].copy()
    else:
        if bg is None:
            return None
        d1 = t_vis[:3] - c_vis[:3]
        d2 = bg[:3] - x_vis[:3]
        den = float(np.dot(d2, d2))
        if den < 1e-6:
            return None
        one_minus = float(np.dot(d1, d2) / den)
        alpha = 1.0 - one_minus
        if not 0.12 <= alpha <= 0.97:
            return None
        tc = (t_vis[:3] - one_minus * bg[:3]) / alpha
        if (tc < -12).any() or (tc > 267).any():
            return None
        tc = np.clip(tc, 0, 255)
    pred = alpha * tc + (1.0 - alpha) * x_vis[:3]
    resid = float(np.sqrt(np.mean((pred - c_vis[:3]) ** 2)))
    # The blend must be a *different* colour from both parents; otherwise C is
    # just an anti-aliased sliver of one of them, not an overlap.
    if np.linalg.norm(c_vis[:3] - t_vis[:3]) < 6.0 or np.linalg.norm(c_vis[:3] - x_vis[:3]) < 6.0:
        return None
    return alpha, tc, resid


def decompose_overlaps(
    labels: np.ndarray,
    fills: dict[int, Fill],
    visible: dict[int, bool],
    curve_params: CurveParams,
    tol: float,
    max_region_frac: float = 0.6,
) -> Decomposition:
    dec = Decomposition()
    edges = adjacency(labels)
    nbrs: dict[int, set[int]] = {}
    for a, b in edges:
        nbrs.setdefault(a, set()).add(b)
        nbrs.setdefault(b, set()).add(a)
    solid_vis = {lab for lab, f in fills.items() if visible.get(lab) and isinstance(f, Solid)}
    if len(solid_vis) < 3:
        return dec

    areas = {lab: int((labels == lab).sum()) for lab in solid_vis}
    border = np.zeros(labels.shape, bool)
    border[0, :] = border[-1, :] = border[:, 0] = border[:, -1] = True
    border_labels = set(np.unique(labels[border]).tolist())
    background = max((lab for lab in solid_vis if lab in border_labels), key=lambda l: areas[l], default=None)
    bg_colour = fills[background].rgba if background is not None and fills[background].rgba[3] > 250 else None
    total = labels.size

    # 1. explain each candidate region by the best (top, under) pair of its neighbours
    explained: dict[int, tuple[int, int, float, np.ndarray]] = {}  # C -> (T, X, alpha, Tc)
    for c in solid_vis:
        if c == background or areas[c] > max_region_frac * total:
            continue
        cand = [n for n in nbrs.get(c, set()) if n in solid_vis and n != background]
        best = None
        for t in cand:
            for x in cand:
                if t == x:
                    continue
                res = _blend(fills[t].rgba, fills[c].rgba, fills[x].rgba, bg_colour)
                if res is None or res[2] > tol:
                    continue
                if best is None or res[2] < best[0]:
                    best = (res[2], t, x, res[0], res[1])
        if best is not None:
            explained[c] = (best[1], best[2], best[3], best[4])

    if not explained:
        return dec

    # An overlap is the part two shapes share, so its outline is theirs: the
    # top's outline where it crosses the shape beneath, the under shape's where
    # it passes under the top, another overlap's where three shapes meet. A
    # region whose outline mostly runs along anything else is not these two
    # shapes' overlap, whatever its colour: a JPEG's whole face (75 000 px)
    # read as a 20 % white eye over a 22-pixel rim sliver whose chroma had
    # rung, was dropped, and was repainted over its own nose and mouth.
    perimeter: dict[int, float] = {}
    for (a, b), (cnt, _g) in edges.items():
        perimeter[a] = perimeter.get(a, 0.0) + cnt
        perimeter[b] = perimeter.get(b, 0.0) + cnt

    def shared(a: int, b: int) -> float:
        return edges.get((min(a, b), max(a, b)), [0.0, 0.0])[0]

    candidates = set(explained)
    for c in sorted(explained):
        t, x = explained[c][0], explained[c][1]
        own = sum(shared(c, n) for n in nbrs.get(c, set()) if n in (t, x) or n in candidates)
        if own < OVERLAP_OUTLINE * perimeter.get(c, 0.0):
            del explained[c]
    if not explained:
        return dec

    # 2. a top shape keeps its absorptions only if the extended outline is simpler
    tops: dict[int, list[int]] = {}
    for c, (t, _x, _a, _tc) in explained.items():
        tops.setdefault(t, []).append(c)
    accepted: dict[int, tuple[int, int, float, np.ndarray]] = {}
    for t, cs in tops.items():
        t_mask = labels == t
        union = t_mask | np.isin(labels, cs)
        if _simpler(union, t_mask, curve_params):
            for c in cs:
                accepted[c] = explained[c]
    if not accepted:
        return dec

    # 3. footprints: T ∪ C for tops, X ∪ C beneath, transitively through removed regions
    members: dict[int, set[int]] = {}

    def add(shape: int, region: int, depth: int = 0) -> None:
        if shape in accepted and depth < 8:  # shape is itself a removed overlap: pass to its owners
            t, x, _a, _tc = accepted[shape]
            add(t, region, depth + 1)
            add(x, region, depth + 1)
            return
        members.setdefault(shape, set()).add(region)

    for c, (t, x, alpha, tc) in accepted.items():
        add(t, c)
        add(x, c)
        dec.removed.add(c)
    def resolve(shape: int, index: int) -> int:
        """Follow a removed region to the shape that owns it (0 = top, 1 = under); cycle-safe."""
        seen = {shape}
        while shape in accepted:
            shape = accepted[shape][index]
            if shape in seen:
                break
            seen.add(shape)
        return shape

    for c, (t, x, alpha, tc) in accepted.items():
        if t not in accepted:
            dec.fills[t] = Solid(np.array([tc[0], tc[1], tc[2], alpha * 255.0]))
            # An opaque region read as a translucent shape over the backdrop:
            # its own area is its colour with the backdrop beneath it.
            if fills[t].rgba[3] >= 250 and bg_colour is not None:
                dec.over_backdrop.add(t)
        top_owner = resolve(t, 0)
        under = resolve(x, 1)
        if top_owner != under and top_owner not in accepted and under not in accepted:
            dec.above.append((top_owner, under))
    for shape, regs in members.items():
        if shape in dec.removed:
            continue
        dec.masks[shape] = (labels == shape) | np.isin(labels, sorted(regs))
    return dec
